clientconn.go

```diff
@@ -1680,16 +1680,35 @@ func (ac *addrConn) tearDown(err error) {
 	ac.updateConnectivityState(connectivity.Shutdown, nil)
 	ac.cancel()
 	ac.curAddr = resolver.Address{}
-	if err == errConnDrain && curTr != nil {
-		// GracefulClose(...) may be executed multiple times when
-		// i) receiving multiple GoAway frames from the server; or
-		// ii) there are concurrent name resolver/Balancer triggered
-		// address removal and GoAway.
-		// We have to unlock and re-lock here because GracefulClose => Close => onClose, which requires locking ac.mu.
-		ac.mu.Unlock()
-		curTr.GracefulClose()
-		ac.mu.Lock()
+
+	if curTr != nil {
+		if err == errConnDrain {
+			// Close the transport gracefully when the subConn is being shutdown.
+			//
+			// GracefulClose() may be executed multiple times if:
+			// - multiple GoAway frames are received from the server
+			// - there are concurrent name resolver or balancer triggered
+			//   address removal and GoAway
+			//
+			// We have to unlock and re-lock here because GracefulClose calls
+			// onClose, which requires locking ac.mu.
+			ac.mu.Unlock()
+			curTr.GracefulClose()
+			ac.mu.Lock()
```

Comment:
`updateAddrs` does the same kind of thing via a `defer` instead. Should we do that here to avoid giving up the lock and then taking it again?  (Also we can remove the local variable `curTr` if we do this.)

Or, maybe just move this stuff to the very end of the function so we don't need to re-take the lock?

Reply:
Moving to the end was simpler and felt cleaner as well. Done.